### PR TITLE
feat: adding live balances to bPool

### PR DESCRIPTION
**Pull request by wei3erHase**

### CODE REVIEW

File: src/contracts/BPool.sol [hunk 1]
```diff
@@ -231,32 +228,45 @@ contract BPool is BBronze, BToken, BMath {
...
     _records[_tokens[index]].index = index;
     _tokens.pop();
-    _records[token] = Record({bound: false, index: 0, denorm: 0, balance: 0});
+    _records[token] = Record({bound: false, index: 0, denorm: 0});
 
     _pushUnderlying(token, msg.sender, bsub(tokenBalance, tokenExitFee));
     _pushUnderlying(token, _factory, tokenExitFee);
   }
 
+  // NOTE: deprecated method, as balances are calculated on-the-fly
```
Comment (EndymionJkb) [inline src/contracts/BPool.sol:237]: Could `gulp` just be removed, as we're doing it continuously/implicitly now? Are we expecting people to call this? Nobody is really using V1, so I don't think we need to worry about compatibility with existing pools.

**Review by EndymionJkb — COMMENTED** [review]
Should document somewhere why you're making this change. It looks like it does what you intend it to do (automatic gulping).

In this simpler contract, we're not handling ETH or making external calls, so there shouldn't be any reentrancy issues (e.g., where people could send tokens and change the balances in the middle of an operation).

File: src/contracts/BPool.sol (hunk 1 above)
Comment (wei3erHase) [reply src/contracts/BPool.sol:237]: we can totally deprecate it 👌 
was thinking that if there's the chance of a contract already developed that does `transfer + gulp`, making `gulp` a no-operation would avoid any revert
but, if there's no current implementation, we can just delete the `gulp` method pointer

File: src/contracts/BPool.sol [hunk 2]
```diff
@@ -359,20 +370,22 @@ contract BPool is BBronze, BToken, BMath {
...
     Record storage outRecord = _records[address(tokenOut)];
 
-    require(tokenAmountOut <= bmul(outRecord.balance, MAX_OUT_RATIO), 'ERR_MAX_OUT_RATIO');
+    uint256 tokenOutBalance = IERC20(tokenOut).balanceOf(address(this));
+
+    require(tokenAmountOut <= bmul(tokenOutBalance, MAX_OUT_RATIO), 'ERR_MAX_OUT_RATIO');
 
     uint256 spotPriceBefore =
-      calcSpotPrice(inRecord.balance, inRecord.denorm, outRecord.balance, outRecord.denorm, _swapFee);
+      calcSpotPrice(tokenOutBalance, inRecord.denorm, tokenOutBalance, outRecord.denorm, _swapFee);
```
Comment (wei3erHase) [inline src/contracts/BPool.sol:378]: 🤦

**Pull request merged**